upf/simplifier.py

```diff
@@ -186,8 +196,115 @@ def walk_lt(self, expression: FNode, args: List[FNode]) -> FNode:
+            value = left.constant_value() - right.constant_value()
+            fnode_constant_values = self._number_to_fnode(value)
+            return fnode_constant_values
+        elif right.is_int_constant() or right.is_real_constant():
+            if right.constant_value() < 0:
+                value = -right.constant_value()
+                fnode_constant_values = self._number_to_fnode(value)
+                return self.manager.Plus(left, fnode_constant_values)
+            else:
+                return self.manager.Minus(left, right)
+        else:
+            return self.manager.Minus(left, right)
+
+    
+    def walk_times(self, expression: FNode, args: List[FNode]) -> FNode:
+        new_args_times: List[FNode] = list()
+        accumulator = 1
+        #divide constant FNode and accumulate their value into accumulator
+        for a in args:
+            if a.is_int_constant() or a.is_real_constant():
+                if a.constant_value() == 0:
+                    return self.manager.Int(0)
+                else:
+                    accumulator = accumulator * a.constant_value()
+            elif a.is_times():
+                for s in a.args():
+                    if s.is_int_constant() or s.is_real_constant():
+                        if s.constant_value() == 0:
+                            return self.manager.Int(0)
+                        else:
+                            accumulator = accumulator * s.constant_value()
+                    else:
+                        new_args_times.append(s)
+            else:
+                new_args_times.append(a)
+        #if accumulator != 1 create it as a constant FNode and then add all the non-constant FNodes found
+        #else return  or all the non-constant FNodes found
+        if accumulator != 1:
+            fnode_acc = self._number_to_fnode(accumulator)
+            return self.manager.Times(*new_args_times, fnode_acc)
+        else: 
+            if len(new_args_times) == 0:
+                return self.manager.Int(1)
+            else:
+                return self.manager.Times(new_args_times)
+    
+            
+        
+    def walk_div(self, expression: FNode, args: List[FNode]) -> FNode:
+        assert len(args) == 2
+        left, right = args
+        value : Union[Fraction, int, float] = 0
+        if left.is_int_constant() and right.is_int_constant():
+            if (left.constant_value() % right.constant_value()) == 0:
+                value = int(left.constant_value() / right.constant_value())
+            else:
+                value = Fraction(left.constant_value(), right.constant_value())
+        elif (left.is_int_constant() or left.is_real_constant()) and (right.is_int_constant() or right.is_real_constant()):
+            assert(right.constant_value() != 0)
+            value = Fraction(left.constant_value(), right.constant_value())
```

Comment:
The Fraction arguments must be integers, but here at least one of the two is a Fraction too.
```suggestion
            value = Fraction(left.constant_value()) / Fraction(right.constant_value())
```

Reply:
Nope, I like @Framba-Luca solution better, there is indeed a `Fraction` constructor with two arguments: https://docs.python.org/3/library/fractions.html#fractions.Fraction

Reply:
You are right!